src/rustAnalyzer.ts

```diff
@@ -0,0 +1,326 @@
+import * as child_process from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { promisify } from 'util';
+
+import * as vs from 'vscode';
+import * as lc from 'vscode-languageclient';
+
+import { WorkspaceProgress } from './extension';
+import { download, fetchRelease } from './net';
+import * as rustup from './rustup';
+import { Observable } from './utils/observable';
+
+const stat = promisify(fs.stat);
+const mkdir = promisify(fs.mkdir);
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+const REQUIRED_COMPONENTS = ['rust-src'];
+
+/** Returns a path where rust-analyzer should be installed. */
+function installDir(): string | undefined {
+  if (process.platform === 'linux' || process.platform === 'darwin') {
+    // Prefer, in this order:
+    // 1. $XDG_BIN_HOME (proposed addition to XDG spec)
+    // 2. $XDG_DATA_HOME/../bin/
+    // 3. $HOME/.local/bin/
+    const { HOME, XDG_DATA_HOME, XDG_BIN_HOME } = process.env;
+    if (XDG_BIN_HOME) {
+      return path.resolve(XDG_BIN_HOME);
+    }
+
+    const baseDir = XDG_DATA_HOME
+      ? path.join(XDG_DATA_HOME, '..')
+      : HOME && path.join(HOME, '.local');
+    return baseDir && path.resolve(path.join(baseDir, 'bin'));
+  } else if (process.platform === 'win32') {
+    // %LocalAppData%\rust-analyzer\
+    const { LocalAppData } = process.env;
+    return (
+      LocalAppData && path.resolve(path.join(LocalAppData, 'rust-analyzer'))
+    );
+  }
+
+  return undefined;
+}
+
+/** Returns a path where persistent data for rust-analyzer should be installed. */
+function metadataDir(): string | undefined {
```

Comment:
@matklad fyi it'd be great to support something like `rust-analyzer --version --tag` or simply include just it in the output to drop the persistent data logic (that bit is copied in spirit from the existing rust-analyzer extension) and query the binary itself

Reply:
The problem here is that `nightly` tag moves. I guess, we can change version string to `rust-analyzer yyyy-mm-dd [nightly]`. Yeah, I if I am not missing something, this would probably be a better way to handle this.

Reply:
Though, @Xanewok, could you explain why we need  to effectively implement our own persistance here? Shouldn't this be handled by VS Code persistance mechanism? Ie, this:

https://github.com/rust-analyzer/rust-analyzer/blob/62730ae30db453f53eb0c37647b5c76058eafca0/editors/code/src/persistent_state.ts

Reply:
> Shouldn't this be handled by VS Code persistance mechanism?

You're absolutely right! What I did at the beginning was to implement everything except the persistence just to get it working; while polishing it I just bolted it on thinking it's implemented similarly (didn't know about the Memento API). It'd be a good idea to drop this altogether :)